src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketMultiBranchLinkActionFactory.java

```diff
@@ -3,24 +3,24 @@
 import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
 import hudson.Extension;
 import hudson.model.Action;
-import hudson.model.Project;
 import hudson.util.FormValidation;
 import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
 
-import javax.annotation.Nonnull;
 import java.util.*;
 
 @Extension
-public class BitbucketLinkActionFactory extends TransientActionFactory<Project> {
+public class BitbucketMultiBranchLinkActionFactory extends TransientActionFactory<WorkflowMultiBranchProject> {
 
-    @Nonnull
     @Override
-    public Collection<? extends Action> createFor(@Nonnull Project target) {
-        if (!(target.getScm() instanceof BitbucketSCM)) {
+    public Collection<? extends Action> createFor(WorkflowMultiBranchProject target) {
+        Optional<BitbucketSCMSource> maybeSource = getBitbucketSCMSource(target);
+        if (!maybeSource.isPresent()) {
             return Collections.emptySet();
         }
-        BitbucketSCM bitbucketSCM = (BitbucketSCM) target.getScm();
-        BitbucketSCM.DescriptorImpl descriptor = (BitbucketSCM.DescriptorImpl) bitbucketSCM.getDescriptor();
+
+        BitbucketSCMSource bitbucketSCM = maybeSource.get();
```

Comment:
This block of code is largely duplicated between each action factory because, unfortunately, there is no unifying interface between `BitbucketSCM`, `BitbucketSCMStep` and `BitbucketSCMSource` in the same way there is for our descriptors. If there was, we could put this in an inherited abstract class, but that would probably deserve it's own PR. Opinions?

Reply:
Re. my earlier comment I think I may end up completely removing this class instead, as the link location is ambiguous in forked workflows across instances. I agree a unifying interface will make such work easier in the future, but I think I prefer that to go to a separate PR- it's a sizeable refactor.

Reply:
Fantastic and I 100% agree with placing that refactor in a seperate PR.

Reply:
So good news- by injecting the form validator and plugin configuration, and using the bitbucket repository instead, we eschew the need for such a refactor- all the information we require is contained therein, and we should use it instead of making new interfaces.